jax/numpy/lax_numpy.py

```diff
@@ -1687,24 +1687,48 @@ def wrapper(*args, **kwargs):
     return asarray(f(*args, **kwargs))
   return wrapper
 
+
+@_wraps(onp.linspace)
 def linspace(start, stop, num=50, endpoint=True, retstep=False, dtype=None,
              axis=0):
+  """Implementation of linspace differentiable in start and stop args."""
   lax._check_user_dtype_supported(dtype, "linspace")
-  try:
-    out = onp.linspace(start, stop, num, endpoint, retstep, dtype, axis)
-    if retstep:
-      return asarray(out[0]), out[1]
-    else:
-      return asarray(out)
-  except TypeError:  # Old versions of onp may lack axis arg.
-    out = onp.linspace(start, stop, num, endpoint, retstep, dtype)
-    if retstep:
-      return moveaxis(asarray(out[0]), 0, axis), out[1]
-    else:
-      return moveaxis(asarray(out), 0, axis)
+  dtype = dtype or lax.dtype(start, stop, float(num))
+  bounds_shape = list(lax.broadcast_shapes(shape(start), shape(stop)))
+  broadcast_start = broadcast_to(start, bounds_shape)
+  axis = len(bounds_shape) if axis == -1 else axis
+  bounds_shape.insert(axis, 1)
+  iota_shape = [1,] * len(bounds_shape)
+  iota_shape[axis] = num
+  delta = (stop - start) / num
+  if endpoint:
+    delta *= num / (num - 1)
+  out = (reshape(broadcast_start, bounds_shape) +
+         reshape(lax.iota(dtype, num), iota_shape) *
+         reshape(delta, bounds_shape))
+  if retstep:
+    return array(out, dtype=dtype), delta
```

Comment:
Better to call `convert_element_type` or similar here, since it's much simpler than calling `array`.

Reply:
fixed